File: integrations/mem0/python/src/purra_mem0/providers.py
"""Optional, bounded SDK providers. No SDK import at module load."""

from __future__ import annotations

import asyncio
import json
import math
import threading
import time
from collections.abc import Awaitable, Callable, Mapping, Sequence
from contextvars import ContextVar
from dataclasses import asdict, dataclass, replace

from purra.contracts import AgentMessage, ModelCompletion, ModelRequest
from purra.model_execution import AgentModelTask, AgentModelTaskRunner
from purra.ports import CancellationSignal

from ._journal import Journal, MemoryError


@dataclass(frozen=True, slots=True)
class MemoryBudget:
    """Durable per-namespace envelope; reservations are never refunded."""

    key: str
    max_llm_calls: int
    max_embedding_calls: int
    max_input_chars: int
    max_output_tokens: int
    result_capacity_target_tokens: int

    def __post_init__(self):
        if not isinstance(self.key, str) or not self.key.strip() or len(self.key) > 512:
            raise ValueError("invalid budget key")
        for name, value in self.limits.items():
            minimum = 1 if name == "result_capacity_target_tokens" else 0
            if type(value) is not int or not minimum <= value <= 2**31 - 1:
                raise ValueError(f"invalid {name}")

    @property
    def limits(self):
        return {k: v for k, v in asdict(self).items() if k != "key"}


@dataclass(frozen=True, slots=True)
class MemoryUsage:
    llm_calls: int
    embedding_calls: int
    input_chars: int
    reserved_output_tokens: int
    reported_input_tokens: int
    reported_output_tokens: int
    unreported_calls: int
    unsettled_calls: int


@dataclass(frozen=True, slots=True)
class EmbeddingResult:
    vectors: Sequence[Sequence[float]]
    input_tokens: int | None = None


@dataclass(frozen=True, slots=True)
class MemoryProviders:
    """Trusted host callbacks with a result-capacity target per call.

    ``complete`` receives messages, the result-capacity target and a signal.
    ``embed`` receives texts and the signal. Both callbacks run on the caller's
    event loop, even though Mem0 itself runs in a worker thread.
    """

    budget: MemoryBudget
    complete: Callable[[Sequence[AgentMessage], int, CancellationSignal], Awaitable[ModelCompletion]]
    embed: Callable[[Sequence[str], CancellationSignal], Awaitable[EmbeddingResult]]

    def __post_init__(self):
        if not isinstance(self.budget, MemoryBudget) or not callable(self.complete) or not callable(self.embed):
            raise TypeError("invalid memory providers")


def run_model(runner: AgentModelTaskRunner, request: ModelRequest):
    """Use the real Run-injected runner, including its authority and accounting.

    Background ingestion should supply its own bounded callback, not a fake Run.
    """
    if not isinstance(runner, AgentModelTaskRunner) or not isinstance(request, ModelRequest):
        raise TypeError("run_model requires a PurrA runner and request")

    async def complete(messages, result_capacity_target_tokens, signal):
        task = AgentModelTask(
            request=request,
            result_capacity_target_tokens=result_capacity_target_tokens,
        )
        return (await runner.complete(messages, task, signal)).completion

    return complete


_current: ContextVar[ProviderExecution | None] = ContextVar("purra_mem0_execution", default=None)


class ProviderExecution:
    def __init__(self, providers, journal: Journal, dimensions, timeout, max_results, max_input):
        self.providers, self.journal, self.dimensions = providers, journal, dimensions
        self.loop = asyncio.get_running_loop()
        self.signal = asyncio.Event()
        self.deadline = time.monotonic() + timeout
        self.max_results, self.max_input = max_results, max_input
        self.operation = None
        self.error = None
        self._lock = threading.Lock()

    def stop(self, code):
        with self._lock:
            if self.error is None:
                self.error = code
                if self.operation is not None:
                    self.journal.provider_error(self.operation, code)
        self.loop.call_soon_threadsafe(self.signal.set)

    def check(self):
        if time.monotonic() >= self.deadline:
            self.stop("memory_timeout")
        if self.error is not None:
            raise MemoryError(self.error)

    def run(self, work):
        token = _current.set(self)
        try:
            self.check()
            result = work()
            self.check()  # Mem0 can swallow provider exceptions in fallback paths.
            return result
        except Exception:
            if self.error is not None:
                raise MemoryError(self.error) from None
            raise
        finally:
            _current.reset(token)

    def invoke(self, kind, values, *, extraction=True):
        self.check()
        cap = (
            self.providers.budget.result_capacity_target_tokens
            if kind == "llm"
            else 0
        )
        chars = sum(len(v.content) for v in values) if kind == "llm" else sum(map(len, values))
        try:
            call_id = self.journal.admit(self.providers.budget.key, self.operation, kind, chars, cap)
        except Exception as error:
            self.stop(error.code if isinstance(error, MemoryError) else "memory_provider_error")
            raise MemoryError(self.error) from None
        input_tokens = generation_tokens = None
        try:
            self.check()
            async def dispatch():
                self.check()
                if kind == "llm":
                    return await self.providers.complete(values, cap, self.signal)
                return await self.providers.embed(values, self.signal)
            result = asyncio.run_coroutine_threadsafe(dispatch(), self.loop).result()
            if kind == "llm":
                if not isinstance(result, ModelCompletion):
                    raise MemoryError("memory_provider_contract")
                if result.usage is not None:
                    input_tokens, generation_tokens = result.usage.input_tokens, result.usage.generation_tokens
                if (result.applied_generation_limit is None or result.finish_reason != "stop"
                        or result.message.tool_calls or result.message.role != "assistant"
                        or result.applied_generation_limit < cap):
                    raise MemoryError("memory_provider_contract")
                content = result.message.content
                if extraction:
                    try:
                        parsed = json.loads(content)
                    except (TypeError, ValueError):
                        raise MemoryError("memory_invalid_extraction")
                    if (not isinstance(parsed, dict) or not isinstance(parsed.get("memory"), list)
                            or len(parsed["memory"]) > self.max_results):
                        raise MemoryError("memory_invalid_extraction")
                    for item in parsed["memory"]:
                        if (not isinstance(item, dict) or not isinstance(item.get("text"), str)
                                or not item["text"].strip() or len(item["text"]) > self.max_input
                                or not isinstance(item.get("entities", []), list)
                                or any(not isinstance(e, str) for e in item.get("entities", []))):
                            raise MemoryError("memory_invalid_extraction")
            else:
                if not isinstance(result, EmbeddingResult):
                    raise MemoryError("memory_provider_contract")
                input_tokens, generation_tokens = result.input_tokens, 0
                if input_tokens is not None and (type(input_tokens) is not int or not 0 <= input_tokens <= 2**31 - 1):
                    input_tokens = None
                    raise MemoryError("memory_provider_contract")
                vectors = [list(v) for v in result.vectors]
                if len(vectors) != len(values) or any(len(v) != self.dimensions or any(
                    isinstance(x, bool) or not isinstance(x, (int, float)) or not math.isfinite(x) for x in v
                ) for v in vectors):
                    raise MemoryError("memory_provider_contract")
                content = vectors
            self.check()
            self.journal.settle(call_id, "complete", input_tokens, generation_tokens)
            return content
        except Exception as error:
            self.journal.settle(call_id, "failed", input_tokens, generation_tokens)
            code = error.code if isinstance(error, MemoryError) else "memory_provider_error"
            self.stop(code)
            raise MemoryError(self.error) from None


def current_execution(journal=None):
    value = _current.get()
    return value if journal is None or value is not None and value.journal is journal else None


def _execution():
    execution = current_execution()
    if execution is None:
        raise MemoryError("memory_provider_unbound")
    return execution


class ManagedMem0Client:
    """SDK handle paired with guarded provider entry points; host owns ``sdk``."""

    def __init__(self, sdk, dimensions):
        self.sdk, self.dimensions = sdk, dimensions

    def __getattr__(self, name):
        return getattr(self.sdk, name)


def create_managed_client(*, config: Mapping, embedding_dims: int):
    """Construct the package-private Mem0 extension with native PurrA providers.

    Only storage/history/custom extraction instructions are accepted. Rerankers,
    graph memory and alternate provider configuration cannot bypass admission.
    Set MEM0_TELEMETRY=false and paths before calling, as for the raw SDK.
    """
    if type(embedding_dims) is not int or not 1 <= embedding_dims <= 65_536:
        raise ValueError("invalid embedding_dims")
    if not isinstance(config, Mapping) or set(config) - {"vector_store", "history_db_path", "custom_instructions"}:
        raise ValueError("managed config accepts only vector_store, history_db_path, custom_instructions")
    if not config.get("vector_store") or not config.get("history_db_path"):
        raise ValueError("explicit vector_store and history_db_path are required")
    vector = config["vector_store"]
    if not isinstance(vector, Mapping) or str(vector.get("provider", "")).lower() == "langchain" or not isinstance(vector.get("config"), Mapping):
        raise ValueError("explicit native vector_store configuration is required")
    values = dict(vector["config"])
    for name in ("embedding_model_dims", "dimension"):
        if name in values and (type(values[name]) is not int or values[name] != embedding_dims):
            raise ValueError("vector store dimension must match embedding_dims")
    if vector.get("provider") == "qdrant":
        values["embedding_model_dims"] = embedding_dims
    from importlib.metadata import version
    if version("mem0ai") != "2.0.19":
        raise RuntimeError("purra-mem0 requires mem0ai 2.0.19")
    from mem0.configs.base import MemoryConfig
    from ._vendor.memory import Memory
    from .direct_providers import DirectLlm, DirectEmbedder
    settings = MemoryConfig(**{**config, "vector_store": {**vector, "config": values}})
    sdk = Memory(settings, llm=DirectLlm(), embedder=DirectEmbedder())
    return ManagedMem0Client(sdk, embedding_dims)

File: integrations/mem0/typescript/src/providers.ts
import { AsyncLocalStorage } from "node:async_hooks";
import type { Message, ModelTaskRunner, ModelTurn } from "purra";
import { Journal, MemoryError } from "./journal.js";
import type { Mem0Client } from "./memory.js";

/** Durable per-namespace envelope. Reservations are never refunded. */
export interface MemoryBudget {
  readonly key: string;
  readonly maxLlmCalls: number;
  readonly maxEmbeddingCalls: number;
  readonly maxInputChars: number;
  readonly maxOutputTokens: number;
  /** Per-call result sizing target. This does not reduce the Provider generation allowance. */
  readonly resultCapacityTargetTokens: number;
}
export interface MemoryUsage {
  readonly llmCalls: number;
  readonly embeddingCalls: number;
  readonly inputChars: number;
  readonly reservedOutputTokens: number;
  readonly reportedInputTokens: number;
  readonly reportedOutputTokens: number;
  readonly unreportedCalls: number;
  readonly unsettledCalls: number;
}
export interface EmbeddingResult { readonly vectors: readonly (readonly number[])[]; readonly inputTokens?: number }
/** Trusted callbacks: preserve the result-capacity target and disable hidden retries. */
export interface MemoryProviders {
  readonly budget: MemoryBudget;
  readonly complete: (messages: readonly Message[], resultCapacityTargetTokens: number, signal: AbortSignal) => Promise<ModelTurn>;
  readonly embed: (texts: readonly string[], signal: AbortSignal) => Promise<EmbeddingResult>;
}

/** Use the real Run-injected runner. Background ingestion must not fabricate a Run. */
export function runModel(runner: ModelTaskRunner): MemoryProviders["complete"] {
  return async (messages, resultCapacityTargetTokens, signal) => (await runner.complete(messages, {
    resultCapacityTargetTokens,
    resultCapacitySource: "workflow_policy",
    signal,
  })).turn;
}

export function providerLimits(providers: MemoryProviders): Record<string, number> {
  if (typeof providers?.complete !== "function" || typeof providers?.embed !== "function") throw new TypeError("invalid memory providers");
  const b = providers.budget;
  if (typeof b?.key !== "string" || !b.key.trim() || [...b.key].length > 512) throw new TypeError("invalid budget key");
  const limits = { max_llm_calls: b.maxLlmCalls, max_embedding_calls: b.maxEmbeddingCalls,
    max_input_chars: b.maxInputChars, max_output_tokens: b.maxOutputTokens,
    result_capacity_target_tokens: b.resultCapacityTargetTokens };
  for (const [name, value] of Object.entries(limits)) {
    if (!Number.isSafeInteger(value) || value < (name === "result_capacity_target_tokens" ? 1 : 0) || value > 2 ** 31 - 1) throw new TypeError(`invalid ${name}`);
  }
  return limits;
}

const current = new AsyncLocalStorage<ProviderExecution>();
export function currentExecution(journal?: Journal): ProviderExecution | undefined {
  const value = current.getStore();
  return journal === undefined || value?.journal === journal ? value : undefined;
}
function execution(): ProviderExecution {
  const value = currentExecution();
  if (!value) throw new MemoryError("memory_provider_unbound");
  return value;
}

export class ProviderExecution {
  readonly controller = new AbortController();
  readonly deadline: number;
  operation: string | undefined;
  error: string | undefined;
  constructor(readonly providers: MemoryProviders, readonly journal: Journal, readonly dimensions: number,
    timeoutMs: number, readonly maxResults: number, readonly maxInput: number) { this.deadline = performance.now() + timeoutMs; }
  stop(code: string): void {
    if (!this.error) {
      this.error = code;
      if (this.operation) this.journal.providerError(this.operation, code);
    }
    this.controller.abort(new MemoryError(this.error));
  }
  check(): void {
    if (performance.now() >= this.deadline) this.stop("memory_timeout");
    if (this.error) throw new MemoryError(this.error);
  }
  run<T>(work: () => Promise<T>): Promise<T> {
    return current.run(this, async () => {
      try {
        this.check();
        const result = await work();
        this.check(); // SDK fallback must not convert a swallowed denial to success.
        return result;
      } catch (error) {
        if (this.error) throw new MemoryError(this.error);
        throw error;
      }
    });
  }
  async invoke(kind: "llm", values: readonly Message[], extraction?: boolean): Promise<string>;
  async invoke(kind: "embedding", values: readonly string[]): Promise<number[][]>;
  async invoke(kind: "llm" | "embedding", values: readonly Message[] | readonly string[], extraction = true): Promise<string | number[][]> {
    this.check();
    const resultTarget = kind === "llm" ? this.providers.budget.resultCapacityTargetTokens : 0;
    const chars = values.reduce<number>((sum, value) => sum + [...(typeof value === "string" ? value : value.content as string)].length, 0);
    let id: string;
    try { id = this.journal.admit(this.providers.budget.key, this.operation, kind, chars, resultTarget); }
    catch (error) { this.stop(error instanceof MemoryError ? error.code : "memory_provider_error"); throw new MemoryError(this.error!); }
    let inputTokens: number | null = null;
    let generationTokens: number | null = null;
    try {
      this.check();
      let content: string | number[][];
      if (kind === "llm") {
        const result = await this.providers.complete(
          values as readonly Message[],
          resultTarget,
          this.controller.signal,
        );
        if (result?.usage) {
          if (!validTokens(result.usage.inputTokens) || (result.usage.generationTokens !== undefined && !validTokens(result.usage.generationTokens))) {
            throw new MemoryError("memory_provider_contract");
          }
          inputTokens = result.usage.inputTokens;
          generationTokens = result.usage.generationTokens ?? null;
        }
        const appliedGenerationLimit = result?.appliedGenerationLimit;
        if (!Number.isSafeInteger(appliedGenerationLimit) || (appliedGenerationLimit as number) < resultTarget
            || result.finishReason !== "stop" || result.message?.role !== "assistant"
            || result.message.toolCalls?.length || typeof result.message.content !== "string"
            || (generationTokens !== null && generationTokens > (appliedGenerationLimit as number))) {
          throw new MemoryError("memory_provider_contract");
        }
        content = result.message.content;
        if (extraction) {
          let parsed;
          try { parsed = JSON.parse(content) as { memory?: unknown }; }
          catch { throw new MemoryError("memory_invalid_extraction"); }
          if (!parsed || !Array.isArray(parsed.memory) || parsed.memory.length > this.maxResults) throw new MemoryError("memory_invalid_extraction");
          for (const item of parsed.memory as { text?: unknown; entities?: unknown }[]) {
            if (!item || typeof item.text !== "string" || !item.text.trim() || [...item.text].length > this.maxInput
                || (item.entities !== undefined && (!Array.isArray(item.entities) || item.entities.some(e => typeof e !== "string")))) {
              throw new MemoryError("memory_invalid_extraction");
            }
          }
        }
      } else {
        const result = await this.providers.embed(values as readonly string[], this.controller.signal);
        if (result?.inputTokens !== undefined && !validTokens(result.inputTokens)) throw new MemoryError("memory_provider_contract");
        inputTokens = result.inputTokens ?? null;
        generationTokens = 0;
        if (!Array.isArray(result.vectors) || result.vectors.length !== values.length || result.vectors.some(v =>
          !Array.isArray(v) || v.length !== this.dimensions || v.some(x => typeof x !== "number" || !Number.isFinite(x)))) {
          throw new MemoryError("memory_provider_contract");
        }
        content = result.vectors.map(v => [...v]);
      }
      this.check();
      this.journal.settle(id, "complete", inputTokens, generationTokens);
      return content;
    } catch (error) {
      this.journal.settle(id, "failed", inputTokens, generationTokens);
      this.stop(error instanceof MemoryError ? error.code : "memory_provider_error");
      throw new MemoryError(this.error!);
    }
  }
}
function validTokens(value: unknown): value is number { return Number.isSafeInteger(value) && (value as number) >= 0 && (value as number) <= 2 ** 31 - 1; }

/** Host owns the SDK handle and its storage resources. */
export class ManagedMem0Client implements Mem0Client {
  constructor(readonly sdk: Mem0Client, readonly dimensions: number) {}
  add(...args: Parameters<Mem0Client["add"]>): Promise<unknown> { return this.sdk.add(...args); }
  get(...args: Parameters<Mem0Client["get"]>): Promise<unknown> { return this.sdk.get(...args); }
  getAll(...args: Parameters<Mem0Client["getAll"]>): Promise<unknown> { return this.sdk.getAll(...args); }
  search(...args: Parameters<Mem0Client["search"]>): Promise<unknown> { return this.sdk.search(...args); }
  update(...args: Parameters<Mem0Client["update"]>): Promise<unknown> { return this.sdk.update(...args); }
  delete(...args: Parameters<Mem0Client["delete"]>): Promise<unknown> { return this.sdk.delete(...args); }
  history(...args: Parameters<Mem0Client["history"]>): Promise<unknown> { return this.sdk.history(...args); }
}

export interface ManagedMem0Config {
  readonly vectorStore: { readonly provider: string; readonly config: Record<string, unknown> };
  readonly historyDbPath: string;
  readonly customInstructions?: string;
}

/** Only storage/history/instructions are accepted; no unmetered reranker or graph provider. */
export async function createManagedClient(options: { config: ManagedMem0Config; embeddingDims: number }): Promise<ManagedMem0Client> {
  const { config, embeddingDims } = options;
  if (!Number.isSafeInteger(embeddingDims) || embeddingDims < 1 || embeddingDims > 65_536) throw new TypeError("invalid embeddingDims");
  if (!config || Object.keys(config).some(k => !["vectorStore", "historyDbPath", "customInstructions"].includes(k))) throw new TypeError("managed config accepts only storage/history/instructions");
  if (!config.vectorStore || typeof config.historyDbPath !== "string" || !config.historyDbPath.trim()) throw new TypeError("explicit vectorStore and historyDbPath are required");
  const { Memory } = await import("mem0ai/oss");
  const sdk = new Memory({ ...config,
    llm: { provider: "langchain", config: { model: { async invoke(messages: { type?: string; getType?: () => string; content: unknown }[]) {
      const roles = { human: "user", ai: "assistant", system: "system" } as const;
      const converted = messages.map(message => {
        const role = roles[(message.type ?? message.getType?.()) as keyof typeof roles];
        if (!role || typeof message.content !== "string") throw new MemoryError("memory_provider_contract");
        return { role, content: message.content };
      });
      return { content: await execution().invoke("llm", converted) };
    } } } },
    embedder: { provider: "langchain", config: { embeddingDims, model: {
      async embedQuery(text: string) { return (await execution().invoke("embedding", [text]))[0]!; },
      async embedDocuments(texts: string[]) { return execution().invoke("embedding", texts); },
    } } },
  } as ConstructorParameters<typeof Memory>[0]);
  return new ManagedMem0Client(sdk, embeddingDims);
}
